### PR TITLE
Thomas/fault tolerant rng state

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

DRAFT

Add support for reloading global random state in a map-style dataset when fault-tolerant training is enabled. 
Not implemented (will be addressed in future PR):
- multiple training dataloaders (combined loader)
- local random state through generators inside the dataset

Blocked by #8891 

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/auto_restart.py [hunk 1]
```diff
@@ -168,6 +172,16 @@ def update(self, generator_name: Optional[str], new_state: IteratorState) -> Non
         state[latest_worker_id] = new_state
         self.latest_worker_id = latest_worker_id
 
+    @property
+    def sampler_states(self) -> Dict[int, Any]:
+        # TODO: add docs
+        # TODO: double check the index here:
+        return {0: self.state[k].sampler_state[0] for k in self.state.keys()}
```
Comment (awaelchli) [inline pytorch_lightning/utilities/auto_restart.py:179]: @tchaton is the index 0 correct here?

Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:179]: I am not entirely sure tbh. We need some test with num_workers > 0.

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -1395,4 +1395,8 @@ def _on_exception(self):
             return
         # save a checkpoint for fault tolerant training. we don't use `log_dir` to minimize the chances of failure.
         file_path = os.path.join(self.default_root_dir, ".pl_auto_save.ckpt")
+
+        # if a previous `pl_auto_save` was saved, delete it.
+        if os.path.exists(file_path):
+            os.remove(file_path)
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:1401]: this should use fsspec / cloud_io in case default_root_dir is not a local storage path

Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:1401]: @tchaton do we need this at all? at this point we know we will save a checkpoint so if a file exists we would overwrite it.

Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:1401]: For the double failure test, it didn't seem to overwrite the file. Might be worth to double check.

File: pytorch_lightning/trainer/supporters.py [hunk 3]
```diff
@@ -161,6 +164,7 @@ def __init__(self, loader: Any, length: Optional[int] = None, state: SharedCycle
         self.loader = loader
         self._loader_iter = None
         self.counter = 0
+        self.state = state
```
Comment (tchaton) [inline pytorch_lightning/trainer/supporters.py:167]: typing ?

Comment (awaelchli) [reply pytorch_lightning/trainer/supporters.py:167]: it is typed as SharedCycleIteratorState

File: pytorch_lightning/utilities/auto_restart.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/auto_restart.py:179]: we only use the sampler_states property in our test, not in the lightning integration. the test works with num_workers and if I change it here, it breaks.

File: pytorch_lightning/trainer/supporters.py [hunk 4]
```diff
@@ -354,80 +364,103 @@ def __init__(self, loaders: Any, mode: str = "min_size"):
...
-        else:
-            # fetch directly from fast forward sampler
-            state.update(dataloader.fast_forward_sampler.state_dict(num_batches_processed))
-        return DataLoaderDict(state)
-
-    def state_dict(self, num_batches_processed: int) -> Dict:
+    def _state_dict_fn(dataloader: DataLoader, iterator: Optional[Iterator], has_completed: int) -> Dict:
+        if isinstance(dataloader, CycleIterator):
+            iterator = dataloader._loader_iter
+        state = getattr(iterator, "state", None) if has_completed else getattr(iterator, "previous_state", None)
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/trainer/supporters.py:370]: updated

File: pytorch_lightning/loops/fit_loop.py [hunk 5]
```diff
@@ -234,3 +240,12 @@ def should_accumulate(self) -> bool:
 
     def teardown(self) -> None:
         self.epoch_loop.teardown()
+
+    def on_save_checkpoint(self) -> Dict:
+        state_dict = super().on_save_checkpoint()
+        state_dict["dataloader_state_dict"] = self.trainer.train_dataloader.state_dict(False)
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/loops/fit_loop.py]: we probably should not set False here everytime in the loop. Because the loop can also save when the iterator has completed.

**Review by tchaton — APPROVED** [review]
Great progress !

File: pytorch_lightning/loops/fit_loop.py [hunk 6]
```diff
@@ -40,6 +40,8 @@ def __init__(self, min_epochs: Optional[int] = None, max_epochs: Optional[int] =
         self.min_epochs = min_epochs
         self.epoch_loop: Optional[TrainingEpochLoop] = None
         self.epoch_progress = Progress()
+        # caches the loaded dataloader state until dataloader objects are available
+        self._dataloader_state_dict: Dict = {}
```
Comment (carmocca) [inline pytorch_lightning/loops/fit_loop.py:44]: Perhaps worth adding the key and value type?

File: pytorch_lightning/loops/fit_loop.py [hunk 7]
```diff
@@ -175,6 +177,10 @@ def on_advance_start(self) -> None:
         if self.current_epoch != 0 and self.trainer._should_reload_dl_epoch:
             self.trainer.reset_train_dataloader(model)
 
+        if self._dataloader_state_dict:
+            self.trainer.train_dataloader.load_state_dict(self._dataloader_state_dict)
+            self._dataloader_state_dict = {}
```
Comment (carmocca) [inline pytorch_lightning/loops/fit_loop.py:182]: Did you consider moving this inside `reset_train_dataloader`?

File: pytorch_lightning/trainer/supporters.py [hunk 8]
```diff
@@ -360,80 +371,106 @@ def __init__(self, loaders: Any, mode: str = "min_size"):
...
+        if state:
+            return DataLoaderDict(**asdict(state))
+        return DataLoaderDict()
```
Comment (carmocca) [inline pytorch_lightning/trainer/supporters.py:380]: IMO `DataLoaderDict` should be removed to avoid pickling it. Instead doing something like we do here:

https://github.com/PyTorchLightning/pytorch-lightning/blob/1e4d8929fb4fe79877fe5996a793a42ceb8cdb4b/pytorch_lightning/trainer/connectors/logger_connector/result.py#L653-L659

File: pytorch_lightning/trainer/supporters.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/supporters.py:377]: Do we need `has_completed`? couldn't we know by checking which one is available and which is `None`?

File: pytorch_lightning/trainer/supporters.py [hunk 9]
```diff
@@ -457,6 +494,7 @@ def _wrap_loaders_max_size_cycle(self) -> Any:
             self.loaders = apply_to_collection(
                 self.loaders, Iterable, CycleIterator, length=length, state=state, wrong_dtype=(Sequence, Mapping)
             )
+            state.reset()
 
             state.reset()
```
Comment (carmocca) [inline pytorch_lightning/trainer/supporters.py:499]: twice?

File: pytorch_lightning/trainer/supporters.py [hunk 10]
```diff
@@ -544,6 +583,7 @@ def request_next_batch(loader_iters: Union[Iterator, Sequence, Mapping]) -> Any:
         Returns
             Any: a collections of batch data
         """
+        return apply_to_collection(loader_iters, Iterator, next)
 
         def next_fn(iterator: Iterator):
             batch = next(iterator)
```
Comment (carmocca) [inline pytorch_lightning/trainer/supporters.py:589]: Unreachable code?

File: pytorch_lightning/trainer/supporters.py [hunk 11]
```diff
@@ -574,6 +614,7 @@ def create_loader_iters(
         Returns
             a collections of iterators
         """
+
```
Comment (carmocca) [inline pytorch_lightning/trainer/supporters.py:617]: unnecessary space added

File: pytorch_lightning/trainer/supporters.py [hunk 12]
```diff
@@ -472,6 +510,7 @@ def __getstate__patch__(*_):
 
         _BaseDataLoaderIter.__getstate__ = __getstate__patch__
         iterator = CombinedLoaderIterator(self.loaders)
+
```
Comment (carmocca) [inline pytorch_lightning/trainer/supporters.py:513]: unnecessary space added

File: pytorch_lightning/utilities/auto_restart.py [hunk 13]
```diff
@@ -227,7 +245,19 @@ def load_state_dict(self, state_dict: Dict[int, Any], latest_worker_id: int, num
         self._cached_state_dict = state_dict
 
     def _state_dict(self) -> Dict[int, Dict[str, Any]]:
-        return {self.worker_id: {"rng_states": {}}}
+        return {self.worker_id: {"rng_states": collect_rng_states()}}
+
+
+def collect_rng_states() -> Dict[str, Any]:
+    """Collect the global random state of :mod:`torch`, :mod:`numpy` and Python."""
+    return {"torch": torch.get_rng_state(), "numpy": np.random.get_state(), "python": python_get_rng_state()}
```
Comment (carmocca) [inline pytorch_lightning/utilities/auto_restart.py:253]: Should we also be saving `torch.cuda.get_rng_state`?

File: tests/loops/test_loops.py [hunk 14]
```diff
@@ -504,7 +504,14 @@ def configure_optimizers_multiple(self):
     assert checkpoint["loops"]["fit_loop"] == expected
 
     trainer.fit_loop.load_state_dict(checkpoint["loops"]["fit_loop"], restart_progress=False)
-    assert trainer.fit_loop.state_dict() == checkpoint["loops"]["fit_loop"]
+    state_dict = trainer.fit_loop.state_dict()
+
+    # need to remove these elements for comparison; comparing with `fit_loop.state_dict()` would require the
+    # fit loop to have an iterator, which is only available during training
+    checkpoint["loops"]["fit_loop"]["state_dict"]["dataloader_state_dict"] = {}
+    state_dict["state_dict"]["dataloader_state_dict"] = {}
```
Comment (carmocca) [inline tests/loops/test_loops.py:512]: Might be simpler to set these to ANY, as the test is only ment to test the progress tracking

```suggestion
    # need to remove these elements for comparison; comparing with `fit_loop.state_dict()` would require the
    # fit loop to have an iterator, which is only available during training
    state_dict["state_dict"] = ANY
```

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/trainer/supporters.py (hunk 12 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/supporters.py:513]:
```suggestion
```

File: pytorch_lightning/trainer/supporters.py (hunk 11 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/supporters.py:616]:
```suggestion
```

File: pytorch_lightning/loops/fit_loop.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/loops/fit_loop.py:44]: Done !

File: pytorch_lightning/loops/fit_loop.py (hunk 7 above)
Comment (awaelchli) [reply pytorch_lightning/loops/fit_loop.py:182]: is that correct? what if we don't want to reset the loader and still just load the state dict? see the if-clause above

for example are in current epoch 0 and want to resume mid epoch
